File: src/sql/postgres/protocol/CopyFail.zig
const CopyFail = @This();

message: Data = .{ .empty = {} },

pub fn decodeInternal(this: *@This(), comptime Container: type, reader: NewReader(Container)) !void {
    _ = try reader.int4();

    const message = try reader.readZ();
    this.* = .{
        .message = message,
    };
}

pub const decode = DecoderWrap(CopyFail, decodeInternal).decode;

pub fn writeInternal(
    this: *@This(),
    comptime Context: type,
    writer: NewWriter(Context),
) !void {
    const message = this.message.slice();
    const count: u32 = @sizeOf((u32)) + message.len + 1;
    const header = [_]u8{
        'f',
    } ++ toBytes(Int32(count));
    try writer.write(&header);
    try writer.string(message);
}

pub const write = WriteWrap(@This(), writeInternal).write;

const std = @import("std");
const Data = @import("../../shared/Data.rust").Data;
const DecoderWrap = @import("./DecoderWrap.rust").DecoderWrap;
const NewReader = @import("./NewReader.rust").NewReader;
const NewWriter = @import("./NewWriter.rust").NewWriter;
const WriteWrap = @import("./WriteWrap.rust").WriteWrap;
const toBytes = std.mem.toBytes;

const int_types = @import("../types/int_types.rust");
const Int32 = int_types.Int32;

File: src/sql/postgres/protocol/CopyFail.rs
use crate::postgres::AnyPostgresError;
use crate::postgres::types::int_types::int32;
use crate::shared::Data;

use super::new_reader::NewReader;
use super::new_writer::NewWriter;

pub struct CopyFail {
    pub message: Data,
}

impl Default for CopyFail {
    fn default() -> Self {
        Self {
            message: Data::Empty,
        }
    }
}

impl CopyFail {
    pub fn decode_internal<Container: super::new_reader::ReaderContext>(
        mut reader: NewReader<Container>,
    ) -> Result<Self, AnyPostgresError> {
        let _ = reader.int4()?;

        let message = reader.read_z()?;
        Ok(Self { message })
    }

    pub fn write_internal<Context: super::new_writer::WriterContext>(
        &self,
        writer: NewWriter<Context>,
    ) -> Result<(), AnyPostgresError> {
        let message = self.message.slice();
        let count: u32 =
            u32::try_from(core::mem::size_of::<u32>() + message.len() + 1).expect("int cast");
        let mut header = [0u8; 5];
        header[0] = b'f';
        // `int32(count)` returns big-endian `[u8; 4]`.
        header[1..5].copy_from_slice(&int32(count));
        writer.write(&header)?;
        writer.string(message)?;
        Ok(())
    }
}
